Index: opensbi-1.3.1-linux-on-litex-vexriscv/platform/litex/vexriscv/objects.mk
===================================================================
--- opensbi-1.3.1-linux-on-litex-vexriscv.orig/platform/litex/vexriscv/objects.mk
+++ opensbi-1.3.1-linux-on-litex-vexriscv/platform/litex/vexriscv/objects.mk
@@ -17,10 +17,18 @@ platform-objs-y += platform.o
 # Command for platform specific "make run"
 platform-runcmd = echo LiteX/VexRiscv SMP
 
-PLATFORM_RISCV_XLEN = 32
-PLATFORM_RISCV_ABI = ilp32
-PLATFORM_RISCV_ISA = rv32ima
-PLATFORM_RISCV_CODE_MODEL = medany
+#PLATFORM_RISCV_XLEN = 32
+#PLATFORM_RISCV_ABI = ilp32
+# openSBI is able to detect if zicsr_zifenci must be
+# added, but without an explicit PLATFORM_RISCV_ISA
+# default is imafdc. So fails with default configuration
+ifeq ($(CC_SUPPORT_ZICSR_ZIFENCEI), y)
+	PLATFORM_RISCV_ISA = rv$(PLATFORM_RISCV_XLEN)ima_zicsr_zifencei
+else
+	PLATFORM_RISCV_ISA = rv$(PLATFORM_RISCV_XLEN)ima
+endif
+#PLATFORM_RISCV_ISA = rv32ima
+#PLATFORM_RISCV_CODE_MODEL = medany
 
 # Blobs to build
 FW_TEXT_START=0x40F00000
