From 72bc8a4743b6a2da18945dae7275ce9d9aa4aed1 Mon Sep 17 00:00:00 2001
From: Antony Pavlov <antonynpavlov@gmail.com>
Date: Thu, 6 Jun 2019 22:44:44 +0300
Subject: [PATCH 3/3] litex_liteeth: add missed netif_stop_queue()

Signed-off-by: Antony Pavlov <antonynpavlov@gmail.com>
---
 drivers/net/ethernet/litex/litex_liteeth.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/net/ethernet/litex/litex_liteeth.c b/drivers/net/ethernet/litex/litex_liteeth.c
index 4fac11f5f7ca..67b2f034f83a 100644
--- a/drivers/net/ethernet/litex/litex_liteeth.c
+++ b/drivers/net/ethernet/litex/litex_liteeth.c
@@ -208,6 +208,8 @@ static int liteeth_stop(struct net_device *netdev)
 {
 	struct liteeth *priv = netdev_priv(netdev);
 
+	netif_stop_queue(netdev);
+
 	del_timer_sync(&priv->poll_timer);
 
 	outreg8(0, priv->base + LITEETH_WRITER_EV_ENABLE);
-- 
2.20.1

