diff --git a/Documentation/devicetree/bindings/soc/litex/litex,soc_controller.yaml b/Documentation/devicetree/bindings/soc/litex/litex,soc_controller.yaml
new file mode 100644
index 000000000..039894265
--- /dev/null
+++ b/Documentation/devicetree/bindings/soc/litex/litex,soc_controller.yaml
@@ -0,0 +1,46 @@
+PDX-License-Identifier: GPL-2.0
+%YAML 1.2
+---
+$id: http://devicetree.org/schemas/soc/litex/litex,soc_controller.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: LiteX SoC Controller driver
+
+description: |
+  This is SoC Controller driver for LiteX SoC Builder.
+  It's purpose is to detect byte order of a SoC, set correct CSR
+  accessors in litex.h header and provide function for other drivers
+  to check if accessors are ready to use.
+
+maintainers:
+  - Karol Gugala <kgugala@antmicro.com>
+  - Mateusz Holenko <mholenko@antmicro.com>
+
+properties:
+  compatible:
+    const: litex,soc_controller
+
+  reg:
+    description: Base address and length of the register space
+
+  status:
+    description: |
+      disables or enables node
+
+    const: "okay"
+
+required:
+  - compatible
+  - reg
+  - status
+
+examples:
+  - |
+
+  soc_ctrl0: soc_controller@f0000000 {
+			compatible = "litex,soc_controller";
+			reg = <0x0 0xf0000000 0x0 0xC>;
+			status = "okay";
+  };
+
+...
diff --git a/drivers/soc/Kconfig b/drivers/soc/Kconfig
index 833e04a78..f00d297f9 100644
--- a/drivers/soc/Kconfig
+++ b/drivers/soc/Kconfig
@@ -6,6 +6,7 @@  source "drivers/soc/atmel/Kconfig"
 source "drivers/soc/bcm/Kconfig"
 source "drivers/soc/fsl/Kconfig"
 source "drivers/soc/imx/Kconfig"
+source "drivers/soc/litex/Kconfig"
 source "drivers/soc/mediatek/Kconfig"
 source "drivers/soc/qcom/Kconfig"
 source "drivers/soc/renesas/Kconfig"
diff --git a/drivers/soc/Makefile b/drivers/soc/Makefile
index 2ec355003..c244d76a7 100644
--- a/drivers/soc/Makefile
+++ b/drivers/soc/Makefile
@@ -12,6 +12,7 @@ obj-y				+= fsl/
 obj-$(CONFIG_ARCH_GEMINI)	+= gemini/
 obj-$(CONFIG_ARCH_MXC)		+= imx/
 obj-$(CONFIG_SOC_XWAY)		+= lantiq/
+obj-$(CONFIG_LITEX_SOC_CONTROLLER) += litex/
 obj-y				+= mediatek/
 obj-y				+= amlogic/
 obj-y				+= qcom/
diff --git a/drivers/soc/litex/Kconfig b/drivers/soc/litex/Kconfig
new file mode 100644
index 000000000..22c78cda0
--- /dev/null
+++ b/drivers/soc/litex/Kconfig
@@ -0,0 +1,14 @@
+# SPDX-License_Identifier: GPL-2.0
+
+menu "Enable LiteX SoC Builder specific drivers"
+
+config LITEX_SOC_CONTROLLER
+	tristate "Enable LiteX SoC Controller driver"
+	help
+	This option enables the SoC Controller Driver which detects
+	endiannes of SoC, sets correct functions in litex.h in runtime
+	and allow other LiteX drivers to initialise
+	All drivers that use functions from litex.h must depend on
+	LITEX_SOC_CONTROLLER
+
+endmenu
diff --git a/drivers/soc/litex/Makefile b/drivers/soc/litex/Makefile
new file mode 100644
index 000000000..98ff7325b
--- /dev/null
+++ b/drivers/soc/litex/Makefile
@@ -0,0 +1,3 @@
+# SPDX-License_Identifier: GPL-2.0
+
+obj-$(CONFIG_LITEX_SOC_CONTROLLER)	+= litex_soc_ctrl.o
diff --git a/drivers/soc/litex/litex_soc_ctrl.c b/drivers/soc/litex/litex_soc_ctrl.c
new file mode 100644
index 000000000..8e5d6833b
--- /dev/null
+++ b/drivers/soc/litex/litex_soc_ctrl.c
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * LiteX SoC Controller Driver
+ *
+ * Copyright (C) 2020 Antmicro
+ *
+ */
+
+#include <linux/litex.h>
+#include <linux/device.h>
+#include <linux/errno.h>
+#include <linux/of.h>
+#include <linux/of_platform.h>
+#include <linux/platform_device.h>
+#include <linux/printk.h>
+#include <linux/module.h>
+#include <linux/errno.h>
+#include <linux/io.h>
+
+#define SCRATCH_REG_OFF		0x04
+#define SCRATCH_REG_VALUE	0x12345678
+#define SCRATCH_LE		0x00000012
+#define SCRATCH_BE		0x12000000
+
+/*
+ * Pointers for accessor functions
+ *
+ * note: or1k and vexriscv kernels use different CONFIG_GENERIC_IOMAP
+ *	setting and it changes the definition of io{read,write}32
+ *	functions so we are checking here to use correct definition
+ */
+#ifndef CONFIG_GENERIC_IOMAP
+unsigned int (*litex_read_reg)(const volatile void __iomem *addr) = NULL;
+void (*litex_write_reg)(u32 val, volatile void __iomem *addr) = NULL;
+#else
+unsigned int (*litex_read_reg)(void __iomem *addr) = NULL;
+void (*litex_write_reg)(u32 val, void __iomem *addr) = NULL;
+#endif
+
+/*
+ * check if accessors are ready to use
+ * returns true if yes - false if not
+ */
+int litex_check_accessors(void)
+{
+	if (litex_read_reg == NULL || litex_write_reg == NULL)
+		return 0;
+	return 1;
+}
+
+struct litex_soc_ctrl_device {
+	void __iomem *base;
+};
+
+/* Check byte order and set correct accessors */
+static int litex_soc_ctrl_check_endianness(void __iomem *reg_addr)
+{
+	u32 reg;
+	u8 le_flag = -1;
+
+	/* Read first part of litex CSR with Little Endian function */
+	reg = ioread32(reg_addr + SCRATCH_REG_OFF);
+
+	if (reg == SCRATCH_LE) {
+		le_flag = 1;
+		litex_read_reg = ioread32;
+		litex_write_reg = iowrite32;
+	} else if (reg == SCRATCH_BE) {
+		le_flag = 0;
+		litex_read_reg = ioread32be;
+		litex_write_reg = iowrite32be;
+	} else {
+		panic("Endianness detection failure! scratch: 0x%x", reg);
+		return -EINVAL;
+	}
+	/* Double check - this time whole CSR */
+	reg = litex_get_reg(reg_addr + SCRATCH_REG_OFF, 4);
+	if (reg != SCRATCH_REG_VALUE) {
+		panic("Endianness detection failure! scratch: 0x%x should be: 0x%x",
+							reg, SCRATCH_REG_VALUE);
+		return -EINVAL;
+	}
+
+	if (le_flag)
+		pr_info("LiteX CSR access configured: Little Endian");
+	else
+		pr_info("LiteX CSR access configured: Big Endian");
+
+	return 0;
+}
+
+static const struct of_device_id litex_soc_ctrl_of_match[] = {
+	{.compatible = "litex,soc_controller"},
+	{},
+};
+
+MODULE_DEVICE_TABLE(of, litex_soc_ctrl_of_match);
+
+static int litex_soc_ctrl_probe(struct platform_device *pdev)
+{
+	struct device *dev;
+	struct device_node *node;
+	const struct of_device_id *id;
+	struct litex_soc_ctrl_device *soc_ctrl_dev;
+	struct resource *res;
+
+	dev = &pdev->dev;
+	node = dev->of_node;
+	if (!node)
+		return -ENODEV;
+
+	id = of_match_node(litex_soc_ctrl_of_match, node);
+	if (!id)
+		return -ENODEV;
+
+	soc_ctrl_dev = devm_kzalloc(dev, sizeof(*soc_ctrl_dev), GFP_KERNEL);
+	if (IS_ERR_OR_NULL(soc_ctrl_dev))
+		return -ENOMEM;
+
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	if (IS_ERR_OR_NULL(res))
+		return -EBUSY;
+
+	soc_ctrl_dev->base = devm_of_iomap(dev, node, 0, &res->end);
+	if (IS_ERR_OR_NULL(soc_ctrl_dev->base))
+		return -EIO;
+
+	return litex_soc_ctrl_check_endianness(soc_ctrl_dev->base);
+}
+
+static int litex_soc_ctrl_remove(struct platform_device *pdev)
+{
+	return 0;
+}
+
+static struct platform_driver litex_soc_ctrl_driver = {
+	.driver = {
+		.name = "litex-soc-controller",
+		.of_match_table = of_match_ptr(litex_soc_ctrl_of_match)
+	},
+	.probe = litex_soc_ctrl_probe,
+	.remove = litex_soc_ctrl_remove
+};
+
+module_platform_driver(litex_soc_ctrl_driver);
+MODULE_DESCRIPTION("LiteX SoC Controller driver");
+MODULE_AUTHOR("Antmicro <www.antmicro.com>");
+MODULE_LICENSE("GPL v2");
diff --git a/include/linux/litex.h b/include/linux/litex.h
new file mode 100644
index 000000000000..ddae454b52e5
--- /dev/null
+++ b/include/linux/litex.h
@@ -0,0 +1,68 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef _LINUX_LITEX_H
+#define _LINUX_LITEX_H
+
+#include <linux/io.h>
+#include <linux/types.h>
+#include <linux/compiler_types.h>
+
+#define LITEX_REG_SIZE             0x4
+#define LITEX_SUBREG_SIZE          0x1
+#define LITEX_SUBREG_SIZE_BIT      (LITEX_SUBREG_SIZE * 8)
+
+// pointers to read and write functions
+// set in soc controller driver
+//
+// note: or1k and vexriscv kernels use different CONFIG_GENERIC_IOMAP
+//	setting and it changes the definition of io{read,write}32
+//	functions so we are checking here to use correct definition
+#ifndef CONFIG_GENERIC_IOMAP
+extern unsigned int (*litex_read_reg)(const volatile void __iomem *addr);
+extern void (*litex_write_reg)(u32 val, volatile void __iomem *addr);
+#else
+extern unsigned int (*litex_read_reg)(void __iomem *addr);
+extern void (*litex_write_reg)(u32 val, void __iomem *addr);
+#endif
+
+// function implemented in
+// drivers/soc/litex/litex_soc_controller.c
+// to check if accessors are ready to use
+// returns true if yes - false if not
+//
+// Important: all drivers that use functions from this header
+// must check at the beginning of their probe()
+// if litex_read_reg and litex_write_reg are initialized
+// and return -EPROBE_DEFER when they are not
+//
+// example:
+// if (!litex_check_accessors())
+//	return -EPROBE_DEFER;
+int litex_check_accessors(void);
+
+// Helper functions for manipulating LiteX registers
+static inline void litex_set_reg(void __iomem *reg, u32 reg_size, u32 val)
+{
+	u32 shifted_data, shift, i;
+
+	for (i = 0; i < reg_size; ++i) {
+		shift = ((reg_size - i - 1) * LITEX_SUBREG_SIZE_BIT);
+		shifted_data = val >> shift;
+		litex_write_reg(shifted_data, reg + (LITEX_REG_SIZE * i));
+	}
+}
+
+static inline u32 litex_get_reg(void __iomem *reg, u32 reg_size)
+{
+	u32 shifted_data, shift, i;
+	u32 result = 0;
+
+	for (i = 0; i < reg_size; ++i) {
+		shifted_data = litex_read_reg(reg + (LITEX_REG_SIZE * i));
+		shift = ((reg_size - i - 1) * LITEX_SUBREG_SIZE_BIT);
+		result |= (shifted_data << shift);
+	}
+
+	return result;
+}
+
+#endif /* _LINUX_LITEX_H */
